Docs: Add automated theme.json reference documentation

mkaz opened the pull request:
## Description

Uses the theme.json schema document to generate a reference guide for settings and styles.

Fixes: #36278


Note: This is a stacked PR on top of #37558 that adds npm-run-all command, that should be merged first.

## How has this been tested?

See a [rendered view of the documentation](https://github.com/WordPress/gutenberg/blob/docs/36278-gen-theme-ref/docs/reference-guides/theme-json-reference.md).

## Types of changes

Documentation update.

- Generates a new reference document for theme.json
- Adds the script to the docs:build process

mkaz committed:
Initial script to generate settings and style for theme.json

mkaz committed:
Add to table of contents and link to how to guide

- Adds a brief paragraph at top of reference guide linking to how to
  guide for examples

- Adds reference guide to table of contents / manifest

carolinan commented:
Not a code review -the rendered view looks great. 
Wishlist: Include what tools `appearanceTools` enables.

mkaz committed:
Add markdownDescription for appearanceTools

To enrich the automated docs this adds a markdownDescription
field for appearanceTools to explain what tools are enabled.

You can see the markdownDescription field used in the tsconfig
schema here: http://json.schemastore.org/tsconfig

mkaz committed:
Add support for markdownDescription in generated docs

mkaz committed:
Update generated docs.

mkaz commented:
@carolinan Good suggestion, I updated to include the tools that are enabled.

@ajlende I added a `markdownDescription` field and use that if present, this mimicks the [tsconfig schema](http://json.schemastore.org/tsconfig) you showed me. This allows for better formatting for the docs. 👍

ajlende commented on schemas/json/theme.json:
```diff
@@ -10,7 +10,8 @@
 				"appearanceTools": {
-					"description": "Setting that enables ui tools.",
+					"description": "Setting that enables enable UI tools.",
+					"markdownDescription": "Setting that enables the following UI tools:\n\n- border: color, radius, style, width\n- color: link\n- spacing: blockGap, margin, padding\n- typography: lineHeight",
```
I think the additional information here can be added to `description` for now. We've been sticking strictly to the draft-04 spec to ensure the schema will work with the widest variety of tools.

`markdownDescription` was added by Microsoft specifically to support Visual Studio Code (as far as I'm aware) and isn't a valid draft-04 property.

I'd like to see a little bit of research on tool usage for theme developers and on how a few more of the tools handle invalid properties before beginning to include them.

mkaz replied:
ok, I just moved the same richer text to the description field. 👍

ajlende reviewed (commented):
Seems like a solid foundation to build upon!

I'd eventually like to see more descriptions for properties included. Things like formatting for color strings are important and aren't included yet.

Examples are another thing that could be useful. Draft 6 added the `examples` property that we could leverage; however, like `markdownDescription` I'd want to do a little research before moving away from draft-04.

mkaz committed:
Dont use markdownDescription yet, move rich description to description field.

mkaz committed:
Remove markdownDescription field

mkaz committed:
Change doc scripts trigger for pre-commit

Update lint-staged to command to run script to generate
theme.json reference prior to checking for changed files.

For details fixing similar issue see PR #37613

ajlende reviewed (approved):
Looks good to me!
